File: BOJ/2110.cpp
#include <algorithm>
#include <functional>
#include <iostream>
#include <vector>

using namespace std;

int binary_search_max(int min, int max, function<bool(int)> f) {
	while (min <= max) {
		int mid = (min + max) >> 1;
		if (f(mid)) {
			min = mid + 1;
		} else {
			max = mid - 1;
		}
	}
	return max;
}

int main(void) {
	int N, C;

	vector<int> houses;
	
	ios::sync_with_stdio(false);
	cin.tie(NULL);

	// init array
	cin >> N >> C;
	houses.resize(N);
	for (int i = 0; i < N; i++) {
		cin >> houses[i];
	}

	// sort houses in ascending order
	sort(houses.begin(), houses.end());

	// get the maximum of minimum distance between routers
	int answer = binary_search_max(1, (houses.back() - houses.front()) / (C - 1), [&](int distance) {
		int n_router = 1;
		int prev = houses[0];
		for (int house : houses) {
			if (house - prev >= distance) {
				prev = house;
				n_router++;
			}
		}
		return n_router >= C;
	});

	// print the answer
	cout << answer;

	return 0;
}

File: BOJ/2110.ts
import * as fs from 'fs';

const inputManager: Generator<string, any, unknown> = (function *() {
  // const stdin = '/dev/stdin';
  const stdin = './input';
  const lines = fs.readFileSync(stdin).toString().split('\n');
  for (const line of lines) yield line;
  return null;
})();
const nextLine = (): string => inputManager.next().value;

// get all inputs
const [N, C] = nextLine().split(' ').map(n => parseInt(n));
const houses: Array<number> = [];
for (let i = 0; i < N; i++) {
  houses.push(parseInt(nextLine()));
}

// return whether the distance is enough to place all routers
function isValidDistance(distance: Number): boolean {
  let nRouter = 1;
  let prev = houses[0];
  for (let i = 1; i < N; i++) {
    if (houses[i] - prev >= distance) {
      prev = houses[i];
      nRouter++;
    }
  }

  return nRouter >= C;
}

// sort houses in ascending order
houses.sort((a, b) => a - b);

// get the maximum of minimum distance between routers
let minDistance = 1;
let maxDistance = Math.floor((houses[N - 1] - houses[0]) / (C - 1));
while (minDistance <= maxDistance) {
  const distance = Math.floor((minDistance + maxDistance) / 2);
  if (isValidDistance(distance)) {
    minDistance = distance + 1;
  } else {
    maxDistance = distance - 1;
  }
}

const answer = minDistance - 1;

// print the answer
console.log(answer);
